tiledb/sm/serialization/fragment_info.cc

```diff
@@ -0,0 +1,494 @@
+    const URI& array_uri,
+    FragmentInfo* fragment_info) {
+  // Get array_schemas_all from capnp
+  if (fragment_info_reader.hasArraySchemasAll()) {
+    auto all_schemas_reader = fragment_info_reader.getArraySchemasAll();
+
+    if (all_schemas_reader.hasEntries()) {
+      auto entries = fragment_info_reader.getArraySchemasAll().getEntries();
+      for (auto array_schema_build : entries) {
+        auto schema{
+            array_schema_from_capnp(array_schema_build.getValue(), array_uri)};
+        fragment_info->array_schemas_all()[array_schema_build.getKey()] =
+            make_shared<ArraySchema>(HERE(), schema);
+      }
+    }
+  }
+
+  // Get single_fragment_info from capnp
+  if (fragment_info_reader.hasFragmentInfo()) {
+    for (auto single_frag_info_reader :
+         fragment_info_reader.getFragmentInfo()) {
+      auto&& [st, single_frag_info] = single_fragment_info_from_capnp(
+          single_frag_info_reader, fragment_info->array_schemas_all());
+      RETURN_NOT_OK(st);
+      fragment_info->single_fragment_info_vec().emplace_back(
+          single_frag_info.value());
+    }
+  }
+
+  // Get uris to vacuum from capnp
+  if (fragment_info_reader.hasToVacuum()) {
+    for (auto uri : fragment_info_reader.getToVacuum()) {
+      fragment_info->to_vacuum().emplace_back(uri.cStr());
+    }
+  }
+
+  // Fill in the rest of values of fragment info class using the above
+  // deserialized values
+  fragment_info->array_uri() = array_uri;
+  fragment_info->unconsolidated_metadata_num() = 0;
+  for (const auto& f : fragment_info->single_fragment_info_vec()) {
+    fragment_info->unconsolidated_metadata_num() +=
+        (uint32_t)!f.has_consolidated_footer();
+  }
+  // TODO: do we need anterior_ndrange_?
+
+  return Status::Ok();
+}
+
+Status fragment_info_to_capnp(
+    const FragmentInfo& fragment_info,
+    capnp::FragmentInfo::Builder* fragment_info_builder,
+    const bool client_side) {
+  // set array_schema_all
+  const auto& array_schemas_all = fragment_info.array_schemas_all();
+  auto array_schemas_all_builder = fragment_info_builder->initArraySchemasAll();
+  auto entries_builder =
+      array_schemas_all_builder.initEntries(array_schemas_all.size());
+  size_t i = 0;
+  for (const auto& schema : array_schemas_all) {
+    auto entry = entries_builder[i++];
+    entry.setKey(schema.first);
+    auto schema_builder = entry.initValue();
+    RETURN_NOT_OK(array_schema_to_capnp(
+        *(schema.second.get()), &schema_builder, client_side));
+  }
+
+  // set single fragment info list
+  const auto& frag_info = fragment_info.single_fragment_info_vec();
+  auto frag_info_builder =
+      fragment_info_builder->initFragmentInfo(frag_info.size());
+  for (size_t i = 0; i < frag_info.size(); i++) {
+    auto single_info_builder = frag_info_builder[i];
+    single_fragment_info_to_capnp(frag_info[i], &single_info_builder);
+  }
+
+  // set fragment uris to vacuum
+  const auto& uris_to_vacuum = fragment_info.to_vacuum();
```

Comment:
same question about the absolute/relative uri as above.

Reply:
Same as above, yes.